File: Mobile/lib/service/favorite_service.dart
import 'package:gowheel_flutterflow_ui/service/storage_service.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

import '../url.dart';

class FavoriteService {
  static final FavoriteService _instance = FavoriteService._internal();
  factory FavoriteService() => _instance;
  FavoriteService._internal();

  TokenService tokenService = TokenService();

  Future<Map<int, int>> getFavorites() async {
    try {
      final token = await tokenService.getToken();
      final response = await http.get(
        Uri.parse('${URL.baseUrl}User/Favorite/GetAllFavorite'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        }
      );

      if (response.statusCode == 200) {
        final jsonResponse = jsonDecode(response.body);
        if (jsonResponse['success'] == true && jsonResponse['data'] != null) {
          final List<dynamic> favorites = jsonResponse['data'];

          return Map.fromEntries(
            favorites.map((item) => MapEntry(
              item['post']['id'] as int,
              item['id'] as int,
            )),
          );
        }
      }
      return {};
    } catch (e) {
      return {};
    }
  }

  Future<bool> addToFavorite(int postId) async {
    try {
      final token = await tokenService.getToken();
      final response = await http.post(
        Uri.parse('${URL.baseUrl}User/Favorite/AddToFavorite'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode({
          'postId': postId,
        }),
      );

      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }

  Future<bool> removeFromFavorite(int id) async {
    try {
      final token = await tokenService.getToken();
      final response = await http.delete(
        Uri.parse('${URL.baseUrl}User/Favorite/RemoveFavorite/${id}'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }
}

File: FE/src/Service/api/FavoriteService.js
// src/services/AmenityService.js
import axios from "axios";
import FavoriteDTO from "../../DTOs/FavoriteDto"; // Sửa tên file nếu cần
const FavoriteService = {
  async getAllFavorite() {
    const token = sessionStorage.getItem("authToken");
    const response = await axios.get(
      `http://localhost:5027/api/User/Favorite/GetAllFavorite`,
      {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      }
    );
    return response.data;
  },

  async addFavorite(Data) {
    const token = sessionStorage.getItem("authToken");
    const favorite = new FavoriteDTO(Data);
        const jsonData = favorite.toJSON();
    const response = await axios.post(
      `http://localhost:5027/api/User/Favorite/AddToFavorite`,
      jsonData,
      {
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json",
        },
      }
    );
    console.log(response);
    if (!response.success) {
        console.log(response);
    }
    return response;
  },

  async DeleteFavorite(id) {
    const token = sessionStorage.getItem("authToken");
    const response = await axios.delete(
      `http://localhost:5027/api/User/Favorite/RemoveFavorite/${id}`,
      {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      }
    );
    if (response.status != 200) {
      const errorText = response;
      throw new Error(
        `${response.status} - ${errorText}`
      );
    }

    return response;
  },
};
export default FavoriteService;
